### PR TITLE
dbeaver/pro#2795 Editors for objects under a path with special symbols don't get restored

### PR DESCRIPTION
**Pull request by serjiokov**
The node path with spec symbol converted for comparison.

### CODE REVIEW

File: plugins/org.jkiss.dbeaver.model/src/org/jkiss/dbeaver/model/navigator/DBNModel.java [hunk 1]
```diff
@@ -365,7 +365,11 @@ private DBNNode findNodeByPath(
 
         DBNNode detectedNode = null;
         for (DBNNode child : children) {
-            if (child.getNodeId().equals(expectedNodePathName)) {
+            String nodeId = child.getNodeId();
+            if (expectedNodePathName.contains(DBNModel.SLASH_ESCAPE_TOKEN)) {
+                expectedNodePathName = expectedNodePathName.replace(DBNModel.SLASH_ESCAPE_TOKEN, "/");
+            }
```
Comment (ShadelessFox) [inline plugins/org.jkiss.dbeaver.model/src/org/jkiss/dbeaver/model/navigator/DBNModel.java]: Move outside the loop

**Review by ShadelessFox — CHANGES_REQUESTED** [review]

File: plugins/org.jkiss.dbeaver.model/src/org/jkiss/dbeaver/model/navigator/DBNModel.java (hunk 1 above)
Comment (ShadelessFox) [reply plugins/org.jkiss.dbeaver.model/src/org/jkiss/dbeaver/model/navigator/DBNModel.java]: To be honest, I believe this should be done somewhere else. Please take a look at where the path gets deserialized.

Comment (ShadelessFox) [reply plugins/org.jkiss.dbeaver.model/src/org/jkiss/dbeaver/model/navigator/DBNModel.java]: Maybe @alexander-skoblikov has concerns about this change as well

Comment (serjiokov) [reply plugins/org.jkiss.dbeaver.model/src/org/jkiss/dbeaver/model/navigator/DBNModel.java]: Generally, the editor "expectedName" has a node path like `%2F...%2F` In other places node path convertation from `%2F to / `leads to a wrong segment identification.

Comment (serjiokov) [reply plugins/org.jkiss.dbeaver.model/src/org/jkiss/dbeaver/model/navigator/DBNModel.java]: The methods `normalize/denormalize` were extracted to Utils.

**Review by ShadelessFox — COMMENTED** [review]
`normailizeNodePath` -> `escapeNodeId`
`deNormalizeNodePath` -> `unescapeNodeId`
What do you think?

File: plugins/org.jkiss.dbeaver.model/src/org/jkiss/dbeaver/model/navigator/DBNUtils.java [hunk 2]
```diff
@@ -269,4 +272,34 @@ public boolean has(String name) {
...
 
+
+    /**
+     * The method decode ASCII to string
+     *
+     * @param nodePath - path
+     * @return - node path object
+     */
+    public static NodePath decodeNodePath(@NotNull NodePath nodePath) {
+        if (!nodePath.toString().contains(DBNModel.PATTERN_ESCAPE_TOKEN)) {
```
Comment (serge-rider) [inline plugins/org.jkiss.dbeaver.model/src/org/jkiss/dbeaver/model/navigator/DBNUtils.java]: nodePath.toString() is also expensive.

Do not check NodePAth, check `path` String variable.

**Review by serge-rider — CHANGES_REQUESTED** [review]

**Comment by Matvey16** [comment]
We have another error if schema name contains `%`
Maybe it can be fixed in this PR too?

File: plugins/org.jkiss.dbeaver.model/src/org/jkiss/dbeaver/model/navigator/DBNModel.java [hunk 3]
```diff
@@ -57,7 +57,7 @@
  * (e.g. TreeViewer sometimes update only first TreeItem corresponding to model certain model object).
  */
 public class DBNModel implements IResourceChangeListener {
-    public static final String SLASH_ESCAPE_TOKEN = "%2F";
+    public static final String PATTERN_ESCAPE_TOKEN = "%"; //$NON-NLS-1$
```
Comment (ShadelessFox) [inline plugins/org.jkiss.dbeaver.model/src/org/jkiss/dbeaver/model/navigator/DBNModel.java]: Please remove it.

File: plugins/org.jkiss.dbeaver.model/src/org/jkiss/dbeaver/model/navigator/DBNModel.java [hunk 4]
```diff
@@ -271,6 +271,9 @@ private static NodePath getNodePath(@NotNull String path) {
                 break;
             }
         }
+        if (path.contains(DBNModel.PATTERN_ESCAPE_TOKEN)) {
```
Comment (ShadelessFox) [inline plugins/org.jkiss.dbeaver.model/src/org/jkiss/dbeaver/model/navigator/DBNModel.java]: It's used here and assumes _something_ about how paths are escaped. Check this within the `decodeNodePath` method and don't introduce new constants - it introduces unnecessary noise.

**Review by ShadelessFox — CHANGES_REQUESTED** [review]

File: plugins/org.jkiss.dbeaver.model/src/org/jkiss/dbeaver/model/navigator/DBNDatabaseNode.java [hunk 5]
```diff
@@ -833,7 +833,7 @@ public String getNodeItemPath() {
             if (pathName.length() > 0) {
                 pathName.insert(0, '/');
             }
-            pathName.insert(0, node.getNodeDisplayName().replace("/", DBNModel.SLASH_ESCAPE_TOKEN));
```
Comment (ShadelessFox) [inline plugins/org.jkiss.dbeaver.model/src/org/jkiss/dbeaver/model/navigator/DBNDatabaseNode.java:836]: Forget to escape here

File: plugins/org.jkiss.dbeaver.model/src/org/jkiss/dbeaver/model/navigator/DBNUtils.java [hunk 6]
```diff
@@ -269,4 +273,32 @@ public boolean has(String name) {
...
+        try {
+            nodePath = URLDecoder.decode(nodePath, StandardCharsets.UTF_8);
+        } catch (Exception e) {
+            // decoding unsuccessful, skip and use original
+        }
+        return nodePath;
+    }
+
+    /**
+     * The method encode symbols('/', ect)
```
Comment (ShadelessFox) [inline plugins/org.jkiss.dbeaver.model/src/org/jkiss/dbeaver/model/navigator/DBNUtils.java]: etc?

File: plugins/org.jkiss.dbeaver.model/src/org/jkiss/dbeaver/model/navigator/DBNUtils.java [hunk 7]
```diff
@@ -269,4 +273,32 @@ public boolean has(String name) {
...
+     *
+     * @param path - path
+     * @return - string path segment
+     */
+    public static String encodeNodePath(@NotNull String path) {
+        if (path.contains("/")) {
+            return URLEncoder.encode(path, StandardCharsets.UTF_8);
```
Comment (ShadelessFox) [inline plugins/org.jkiss.dbeaver.model/src/org/jkiss/dbeaver/model/navigator/DBNUtils.java]: You **must** escape `%` symbols as well.

**Review by ShadelessFox — CHANGES_REQUESTED** [review]

File: plugins/org.jkiss.dbeaver.model/src/org/jkiss/dbeaver/model/navigator/DBNDatabaseNode.java (hunk 5 above)
Comment (serjiokov) [reply plugins/org.jkiss.dbeaver.model/src/org/jkiss/dbeaver/model/navigator/DBNDatabaseNode.java:836]: +

File: plugins/org.jkiss.dbeaver.model/src/org/jkiss/dbeaver/model/navigator/DBNUtils.java (hunk 7 above)
Comment (serjiokov) [reply plugins/org.jkiss.dbeaver.model/src/org/jkiss/dbeaver/model/navigator/DBNUtils.java]: +

File: plugins/org.jkiss.dbeaver.model/src/org/jkiss/dbeaver/model/navigator/DBNUtils.java [hunk 8]
```diff
@@ -39,6 +40,9 @@
 import org.jkiss.utils.ArrayUtils;
 import org.jkiss.utils.CommonUtils;
 
+import java.net.URLDecoder;
+import java.net.URLEncoder;
+import java.nio.charset.StandardCharsets;
```
Comment (ShadelessFox) [inline plugins/org.jkiss.dbeaver.model/src/org/jkiss/dbeaver/model/navigator/DBNUtils.java]: Redundant imports

File: plugins/org.jkiss.dbeaver.model/src/org/jkiss/dbeaver/model/navigator/DBNUtils.java [hunk 9]
```diff
@@ -29,6 +29,7 @@
 import org.jkiss.dbeaver.model.app.DBPProject;
 import org.jkiss.dbeaver.model.exec.DBCExecutionContext;
 import org.jkiss.dbeaver.model.exec.DBCExecutionContextDefaults;
+import org.jkiss.dbeaver.model.navigator.DBNModel.NodePath;
```
Comment (ShadelessFox) [inline plugins/org.jkiss.dbeaver.model/src/org/jkiss/dbeaver/model/navigator/DBNUtils.java]: Redundant import

File: plugins/org.jkiss.dbeaver.model/src/org/jkiss/dbeaver/model/navigator/DBNModel.java [hunk 10]
```diff
@@ -327,7 +328,8 @@ public DBNNode getNodeByPath(
             log.debug("Project node not found");
             return null;
         }
-        final NodePath nodePath = getNodePath(path);
+        NodePath nodePath = getNodePath(path);
+
```
Comment (ShadelessFox) [inline plugins/org.jkiss.dbeaver.model/src/org/jkiss/dbeaver/model/navigator/DBNModel.java:332]: Redundant change

File: plugins/org.jkiss.dbeaver.model/src/org/jkiss/dbeaver/model/navigator/DBNLegacyUtils.java [hunk 11]
```diff
@@ -139,7 +139,8 @@ private static DBNNode legacyFindNodeByPath(
 
         final List<String> pathItems = nodePath.pathItems;
         for (int i = firstItem, itemsSize = pathItems.size(); i < itemsSize; i++) {
-            String item = pathItems.get(i).replace(DBNModel.SLASH_ESCAPE_TOKEN, "/");
+            //String item = pathItems.get(i).replace(DBNModel.SLASH_ESCAPE_TOKEN, "/");
```
Comment (ShadelessFox) [inline plugins/org.jkiss.dbeaver.model/src/org/jkiss/dbeaver/model/navigator/DBNLegacyUtils.java]: Commented-out code

**Comment by Matvey16** [comment]
Verified

**Pull request merged**